test/hotspot/jtreg/compiler/lib/ir_framework/test/AbstractTest.java

```diff
@@ -151,8 +165,12 @@ protected void compileMethod(DeclaredTest test) {
         checkCompilationLevel(test);
     }
 
-    private void enqueueMethodForCompilation(DeclaredTest test) {
+    private void enqueueMethodForCompilation(DeclaredTest test) throws MethodNotCompilableException {
         final Method testMethod = test.getTestMethod();
+        final boolean isCompilable = WHITE_BOX.isMethodCompilable(testMethod, test.getCompLevel().getValue(), false);
+        if (!isCompilable && test.isAllowNotCompilable()) {
+            throw new MethodNotCompilableException();
+        }
         TestRun.check(WHITE_BOX.isMethodCompilable(testMethod, test.getCompLevel().getValue(), false),
                       "Method " + testMethod + " not compilable (anymore) at level " + test.getCompLevel());
```

Comment:
You now check `isMethodCompilable()` twice. You could move the second check inside the `catch` block for `MethodNotCompilableException`. Then you can use this exception whenever there is a method that is not compilable (as the class name suggests) and then only do the separation between allowed and disallowed cases there.